### ISSUE TITLE
Config file example ?

### ISSUE THREAD

**Issue by Adirelle** [gerbera/gerbera#209]
After compiling and installing gerbera, I tried to "touch /etc/gerbera/config.xml", as told in the doc, and it did not work. I had to google a bit to find an example configuratrion file. I am not fond of build an XML file from its XSD schema. Would it be possible to provide a default or example configuration file to help newcomers ?

**Comment by v00d00** [comment]
Gerbera will generate you a default config on first run?

**Comment by Adirelle** [comment]
That would be cool but maybe too much. However, a default configuration could be generated on build (like the systemd service file), or at least, an example could sit somewhere in the source, with most options provided with their default values.

I had to take http://p.cweiske.de/433 as a starting point.

**Comment by v00d00** [comment]
I mean, the code already does generate you an example config on first run (or if it cant fine a config file).

I guess we could generate one at install time, but I dont see any advantage.

**Comment by Adirelle** [comment]
Oh... This was what I supposed at first. But as I followed the instructions, I created an empty configuration file and gerbera complained it was invalid.

**Comment by elmodaddyb** [comment]
Hello, 

As you noticed the systemd readme is a bit misleading with the `touch config.xml` example. 

That’s my doing 👎 my bad translation of **make sure it exists**

I will be writing up some ui docs soon and will clean it up too. 

If you make sure your `gerbera` runtime user has write access to the path, Gerbera will auto-generate the config for you, if it doesn’t exist.

**Comment by Adirelle** [comment]
Ehh. I'd prefer the service not to have write access to its configuration file unless it is absolutely necessary. Follwoing that idea, maybe Gerbera could have an explicit flag (say -C) to send a default configuration to stdout. This would allow to create it without giving the daemon too much privileges. E.g.

```bash
# Generate the default configuration
sudo mkdir -m 0755 -p /etc/gerbera
gebera -C | sudo tee /etc/gerbera/config.xml
```

**Comment by elmodaddyb** [comment]
Hello, I took a look at the Gerbera code and it is reasonable to make a utility method which generates a default configuration file.  This actually happens today, but the file is written out to file-system instead of standard-out.

Is the feature still desirable?  I will pull something together by leveraging:

```
String ConfigManager::createDefaultConfig(String userhome)
```

1. No more default generation of `config.xml` (_this might be an issue for quick start_)
2. Runtime option to generate the `config.xml`

What are your thoughts?

**Comment by elmodaddyb** [comment]
The **SHOUTcast** capability was using Boost **filesystem** library to unit-test the configuration generation of the new shoutcast config.  This was due in part to the **embedded** file creation in the `ConfigManager::createDefaultConfig(...)` method.

To that end, I think I will tackle this refactor to accomplish 2 things at once.

1. Continue to generate default `config.xml`
2. Abstract away the generation vs. the writing
3. Use **generation** for testability of shoutcast, etc.
4. Use **writing** for standard defaults.

Let me know what you think.

Thanks,

Eamonn

**Comment by elmodaddyb** [comment]
I started working on a new class `ConfigGenerator` which is capable to provide a generated `config.xml` to stdout using some command line switch.  This capability is needed in-turn to test out the SHOUTcast capability.

> Keep watch for updates.

**Comment by elmodaddyb** [comment]
Hello All,

I made good progress on creation of a new `ConfigGenerator` class which generates the XML without writing to the filesystem.  Generation is done and all unit tests pass.

There were a few items to resolve.

1. `config2.xsd` is missing `AppleTrailers` and `ffmpegthumbnailer` specifications
2. System no longer supports `YouTube` XML
3. System no longer supports SQLite3 Backup capability

Any thoughts on **removing** appletrailers?  I plan to cleanup YouTube & SQLite Backup accordingly.

Thanks,

Eamonn

**Comment by v00d00** [comment]
I think _appletrailers_ could work still, there APIs are still live.

**Comment by elmodaddyb** [comment]
OK, will keep _AppleTrailers_.  Here is a status of the work so far.

* **DONE** - Create new `ConfigGenerator` for XML generation
* **DONE** - Wireup new server option `./gerbera --create-config`
* **DONE** - TDD all changes
* **PENDING** - Update `ConfigManager` to use `ConfigGenerator` for default file
* **PENDING** - Run `ConfigGenerator` XML through `ConfigManager.validate()`

> When all complete, `ConfigGenerator` will be the source of truth for: 
> - file generation 
> - console output

_...Then onto SHOUTcast!!!_

**Comment by elmodaddyb** [comment]
Hello All,

The use of `ConfigGenerator` to create a `config.xml` file for Gerbera is complete.  The system now supports:

1. Output to _stdout_ default **config.xml** using command `./gerbera --create-config`
2. Auto-generate **$HOME/.config/gerbera/config.xml** using `ConfigGenerator` on first startup.

The tests coverage is adequate and covers code executed in:
* `config_manager.cc`
* `config_generator.cc`
* `main.cc`

All that remains is code cleanup and integration testing with the new file.  Keep a look-out for a pull request soon!

**Comment by elmodaddyb** [comment]
I removed the automatic generation...instead pointing the user to generate the file manually.  This simplifies the code a bit and removes write activity of the gerbera process.

```
INFO: Gerbera UPnP Server version 1.2.0_alpha - http://gerbera.io/
INFO: ===============================================================================
INFO: Gerbera is free software, covered by the GNU General Public License version 2
INFO: Copyright 2016-2018 Gerbera Contributors.
INFO: Gerbera is based on MediaTomb: Copyright 2005-2010 Gena Batsyan, Sergey Bostandzhyan, Leonhard Wimmer.
INFO: ===============================================================================
ERROR: 
The server configuration file could not be found in ~/.config/gerbera
Gerbera could not find a default configuration file.
Try specifying an alternative configuration file on the command line.
For a list of options run: gerbera -h
```

Let me know any language suggestions with the error message.

Thanks,

Eamonn

**Comment by awissu** [comment]
These changes will not be accepted in distributions.
gerbera --create-config | tee /var/lib/gerbera/config.xml
chown gerbera:gerbera  /var/lib/gerbera/config.xml
systemctl cat gerbera.service
```
[Unit]
Description=UPnP Media Server
After=local-fs.target network.target
Wants=mysql.target

[Service]
EnvironmentFile=/etc/gerbera.conf
PIDFile=/var/run/gerbera.pid
User=gerbera
Group=gerbera
UMask=0007
ExecStart=/usr/bin/gerbera -l /var/log/gerbera/gerbera.log -m /var/lib --create-config /var/lib/gerbera -f gerbera
WorkingDirectory=/var/lib/gerbera
Restart=on-failure
RestartSec=5

[Install]
WantedBy=multi-user.target
```
cat cat /var/log/gerbera/gerbera.log
```
2018-05-03 18:07:02    INFO: Gerbera UPnP Server version 1.2.0 - http://gerbera.io/
2018-05-03 18:07:02    INFO: ===============================================================================
2018-05-03 18:07:02    INFO: Gerbera is free software, covered by the GNU General Public License version 2
2018-05-03 18:07:02    INFO: Copyright 2016-2018 Gerbera Contributors.
2018-05-03 18:07:02    INFO: Gerbera is based on MediaTomb: Copyright 2005-2010 Gena Batsyan, Sergey Bostandzhyan, Leonhard Wimmer.
2018-05-03 18:07:02    INFO: ===============================================================================
2018-05-03 18:07:02    INFO: Loading configuration from: /var/lib/gerbera/config.xml
2018-05-03 18:07:02    INFO: Checking configuration...
2018-05-03 18:07:02   ERROR: Permission denied: /root/.config/gerbera (errno: 130)
```

**Comment by elmodaddyb** [comment]
@awissu - can you help me understand?  Externalizing config generation is not suitable for distribution use?

It is unclear what is wrong?

Do you have any insight to help.

**Comment by elmodaddyb** [comment]
@awissu - Sorry for the confusion.  I understand what has happened.

When the system runs `--create-config` it defaulted the `<home></home>` value to the `/root/.config/gerbera` path.
[http://docs.gerbera.io/en/latest/config-overview.html#generating-configuration](http://docs.gerbera.io/en/latest/config-overview.html#generating-configuration)

I added a note in the documentation, but I see how this is hidden and misleading.

> Any ideas on a more intelligent default?

**Comment by elmodaddyb** [comment]
I did some digging in the source code.  It seems like we could completely eliminate the `<home>` default logic from the system.  However, this would **REQUIRE** `-c config.xml` option to be given always.

> Feedback welcome...

**Comment by pamapa** [comment]
command line handling has been fix on master starting v1.5.0
